edb/pgsql/deltafts.py

```diff
@@ -0,0 +1,471 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from typing import *
+
+from edb import errors
+
+from edb.schema import indexes as s_indexes
+from edb.schema import types as s_types
+from edb.schema import expr as s_expr
+from edb.schema import schema as s_schema
+from edb.schema import delta as sd
+
+from edb.ir import ast as irast
+
+from edb.edgeql import compiler as qlcompiler
+
+from . import common
+from . import dbops
+from . import compiler
+from . import codegen
+from . import types
+from . import ast as pgast
+
+from .common import qname as q
+from .common import quote_literal as ql
+from .common import quote_ident as qi
+from .compiler import astutils
+
+
+def create_fts_index(
+    index: s_indexes.Index,
+    root_code: str,
+    index_expr: irast.Set,
+    predicate_src: Optional[str],
+    sql_kwarg_exprs: Dict[str, str],
+    options: qlcompiler.CompilerOptions,
+    schema: s_schema.Schema,
+    context: sd.CommandContext,
+) -> dbops.Command:
+    subject = index.get_subject(schema)
+    assert isinstance(subject, s_indexes.IndexableSubject)
+
+    effective, has_overridden = s_indexes.get_effective_fts_index(
+        subject, schema
+    )
+
+    if index != effective:
+        return dbops.CommandGroup()
+
+    # When creating an index on a child that already has an fts index
+    # inherited from the parent, we don't need to create the index, but just
+    # update the populating expressions.
+    if has_overridden:
+        return _refresh_fts_document(index, options, schema, context)
+    else:
+        return _create_fts_document(
+            index,
+            root_code,
+            index_expr,
+            predicate_src,
+            sql_kwarg_exprs,
+            schema,
+            context,
+        )
+
+
+def delete_fts_index(
+    index: s_indexes.Index,
+    drop_index: dbops.Command,
+    options: qlcompiler.CompilerOptions,
+    schema: s_schema.Schema,
+    context: sd.CommandContext,
+) -> dbops.Command:
+    subject = index.get_subject(schema)
+    assert isinstance(subject, s_indexes.IndexableSubject)
+
+    effective, _ = s_indexes.get_effective_fts_index(subject, schema)
+
+    if not effective:
+        return _delete_fts_document(index, drop_index, schema, context)
+    else:
+        # effective index remains: don't drop the fts document
+
+        effective_subject = effective.get_subject(schema)
+        is_eff_on_direct_parent = effective_subject in subject.get_bases(
+            schema
+        ).objects(schema)
+
+        if is_eff_on_direct_parent:
```

Comment:
```
type GrandParent {
   index fts::textsearch;
}
type Parent {
   index fts::textsearch;
}
type Child {
}

alter GrandParent {
   drop index fts::textsearch;
}
```

This will also drop the index that is inherited into the Child, but it will not change Child's effective index.